test-integration/reference-book.coffee

```diff
@@ -9,68 +10,111 @@ SECTIONS_TO_TEST = 10
       @waitClick(css: '.menu-toggle')
 
-      doneLoading = =>
-        # Wait until the modal closes after clicking the date
-        @driver.isElementPresent(css: '.loadable.is-loading, .loading-exercise').then (isPresent) -> not isPresent
-
-      for i in [1..SECTIONS_TO_TEST]
-        # Selenium sometimes keeps pressing the same next button (doneLoading doesn't seem to work 100%)
-        @driver.findElement(css: 'a.nav.next').getAttribute('href').then (oldHref) =>
-          @addTimeout(3)
+    checkEachPage = =>
+      hrefToCheck = null
+      nextNav = null
+      # Selenium sometimes keeps pressing the same next button (doneLoading doesn't seem to work 100%)
+      @driver.wait(doneLoading)
+      @driver.findElement(css: 'a.page-navigation.next')
+        .then (element) ->
+          nextNav = element
+          element.getAttribute('href')
+        .then (oldHref) ->
           console.log '----------------'
-          # console.log 'old ashkd', oldHref
-          @driver.findElement(css: 'a.nav.next').click()
-
-          @driver.wait(doneLoading)
-          # @driver.findElement(css: 'a.nav.next').getAttribute('href').then (foo) =>
-          #   console.log 'new ashkd', foo
-
-          checkPageChanged = =>
-            @driver.findElement(css: 'a.nav.next').getAttribute('href').then (newHref) =>
-              newHref isnt oldHref
-
-          ifPageDidntChange = =>
-            @addTimeout(3)
-            console.log 'Page did not change. reclicking.'
-            @driver.findElement(css: 'a.nav.next').click()
-            @driver.wait(doneLoading)
-
-          @driver.wait(checkPageChanged, 1000).then(null, ifPageDidntChange)
-
-
-        @driver.getCurrentUrl().then (pageUrl) =>
-          @addTimeout(3)
-          @driver.findElements(css: '[data-type="exercise"] .question').then (elements) =>
-            if elements.length
-              console.log "Found #{elements.length} exercises"
-
-          @driver.findElements(css: '.reference-book-missing-exercise').then (elements) =>
-            if elements.length > 0
-              console.log "Found #{elements.length} missing exercises in #{pageUrl}"
-              # msg = []
-              # for el in elements
-              #   el.getAttribute('data-exercise-url').then (url) ->
-              #     msg.push(url)
-              #
-              # @driver.getCurrentUrl().then =>
-              #   console.log "Found #{elements.length} missing exercises in #{pageUrl}: #{JSON.stringify(msg)}"
-
-    @driver.get("#{SERVER_URL}books/2")
-    @injectErrorLogging()
-    checkForMissingExercises()
+          console.info("Next href is #{oldHref}. Clicking next.")
+          hrefToCheck = oldHref
+          nextNav.click()
+
+      @driver.wait(doneLoading)
+      checkPageChanged(hrefToCheck)
+
+    doneLoading = =>
+      # sleep to make sure exercises have a chance to start loading before checking if page is indeed done loading
+      @sleep(100)
```

Comment:
Instead of sleeping for a fixed period of time (Sauce and BrowserStack may take longer so it's nice to change this globally) could you wait for an element to either be present or to not be present?